src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -263,124 +293,91 @@ public void visitParenthesizedPattern(JCParenthesizedPattern tree) {
-            } else if (nested.type.isReference() && nested.hasTag(Tag.RECORDPATTERN)) {
-                extraTest = makeBinary(Tag.NE, make.Ident(nestedTemp), makeNull());
+                nestedBinding = nestedDesugared.primaryPattern();
+                allowNull = false;
+            } else {
+                nestedBinding = (JCBindingPattern) nestedPattern;
+                allowNull = true;
             }
-            if (extraTest != null) {
-                extracted = makeBinary(Tag.AND, extraTest, extracted);
+            JCMethodInvocation componentAccessor =
+                    make.App(make.Select(convert(make.Ident(recordBinding), recordBinding.type), //TODO - cast needed????
+                             component.accessor));
+            if (deconstructorCalls == null) {
+                deconstructorCalls = Collections.newSetFromMap(new IdentityHashMap<>());
             }
-            LetExpr getAndRun = make.LetExpr(nestedTempVar, extracted);
-            getAndRun.needsCond = true;
-            getAndRun.setType(syms.booleanType);
-            if (test == null) {
-                test = getAndRun;
+            deconstructorCalls.add(componentAccessor);
+            JCExpression accessedComponentValue =
+                    convert(componentAccessor, componentType);
+            JCInstanceOf firstLevelCheck = (JCInstanceOf) make.TypeTest(accessedComponentValue, nestedBinding).setType(syms.booleanType);
+            //TODO: verify deep/complex nesting with nulls
+            firstLevelCheck.allowNull = allowNull;
+            if (firstLevelChecks == null) {
+                firstLevelChecks = firstLevelCheck;
             } else {
-                test = makeBinary(Tag.AND, test, getAndRun);
+                firstLevelChecks = mergeConditions(firstLevelChecks, firstLevelCheck);
             }
             components = components.tail;
             nestedFullComponentTypes = nestedFullComponentTypes.tail;
             nestedPatterns = nestedPatterns.tail;
         }
-
-        if (tree.var != null) {
-            BindingSymbol binding = (BindingSymbol) tree.var.sym;
-            Type castTargetType = principalType(tree);
-            VarSymbol bindingVar = bindingContext.bindingDeclared(binding);
-
-            JCAssign fakeInit =
-                    (JCAssign) make.at(TreeInfo.getStartPos(tree))
-                                   .Assign(make.Ident(bindingVar),
-                                           convert(make.Ident(currentValue), castTargetType))
-                                   .setType(bindingVar.erasure(types));
-            LetExpr nestedLE = make.LetExpr(List.of(make.Exec(fakeInit)),
-                                            make.Literal(true));
-            nestedLE.needsCond = true;
-            nestedLE.setType(syms.booleanType);
-            test = test != null ? makeBinary(Tag.AND, test, nestedLE) : nestedLE;
-        }
-
         Assert.check(components.isEmpty() == nestedPatterns.isEmpty());
```

Comment:
is it possible for this assertion not to hold under some conditions? can it be removed?